NPE when clicking properties on directory in SFTP mount

## Issue explanation

NPE when clicking properties on directory in SFTP mount

## Exception
* __App Name:__ Amaze 文件管理器
* __Package:__ com.amaze.filemanager
* __Version:__ 3.6.1
* __User Action:__ UI Error
* __Request:__ Application crash
* __OS:__ Linux Android 10 - 29
* __Device:__ HWELE
* __Model:__ ELE-AL00
* __Product:__ ELE-AL00
<details><summary><b>Crash log </b></summary><p>

```
java.lang.RuntimeException: An error occurred while executing doInBackground()
	at android.os.AsyncTask$AsyncFutureTask.done(AsyncTask.java:429)
	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
	at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
	at java.util.concurrent.FutureTask.run(FutureTask.java:271)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:929)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.Boolean.booleanValue()' on a null object reference
	at com.amaze.filemanager.filesystem.HybridFile.isDirectory(HybridFile.java:463)
	at com.amaze.filemanager.asynchronous.asynctasks.CountItemsOrAndSizeTask.doInBackground(CountItemsOrAndSizeTask.java:57)
	at com.amaze.filemanager.asynchronous.asynctasks.CountItemsOrAndSizeTask.doInBackground(CountItemsOrAndSizeTask.java:37)
	at android.os.AsyncTask$3.call(AsyncTask.java:389)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	... 3 more

```
</details>
<hr>

Upon further testing, this always fails when clicking property of any directory.

<details><summary><b>Crash log </b></summary><p>

```
java.lang.RuntimeException: An error occurred while executing doInBackground()
	at android.os.AsyncTask$AsyncFutureTask.done(AsyncTask.java:429)
	at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
	at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
	at java.util.concurrent.FutureTask.run(FutureTask.java:271)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:929)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'long java.lang.Long.longValue()' on a null object reference
	at com.amaze.filemanager.filesystem.HybridFile.getTotal(HybridFile.java:692)
	at com.amaze.filemanager.filesystem.files.FileUtils.getSpaces(FileUtils.java:550)
	at com.amaze.filemanager.asynchronous.asynctasks.LoadFolderSpaceDataTask.doInBackground(LoadFolderSpaceDataTask.java:85)
	at com.amaze.filemanager.asynchronous.asynctasks.LoadFolderSpaceDataTask.doInBackground(LoadFolderSpaceDataTask.java:53)
	at android.os.AsyncTask$3.call(AsyncTask.java:389)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	... 3 more

```
</details>

I've also noticed that when I click into any directory the content is always empty, and upon refreshing a toast would pop up saying something like `Can't read directory: ssh://$address/ - null` (and the title address is `ssh://$address//dir` instead of `ssh://$address/dir` somehow, and if I then use the back gesture the app would crash. Should I file another bug report, or is this related in some way?

<details><summary><b>Crash log </b></summary><p>

```
java.lang.NullPointerException: The callable returned a null value
	at io.reactivex.internal.functions.ObjectHelper.requireNonNull(ObjectHelper.java:39)
	at io.reactivex.internal.operators.single.SingleFromCallable.subscribeActual(SingleFromCallable.java:44)
	at io.reactivex.Single.subscribe(Single.java:3603)
	at io.reactivex.internal.operators.single.SingleSubscribeOn$SubscribeOnObserver.run(SingleSubscribeOn.java:89)
	at io.reactivex.Scheduler$DisposeTask.run(Scheduler.java:578)
	at io.reactivex.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:66)
	at io.reactivex.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:57)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:929)

```
</details>

--- a/app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/GenerateHashesTask.java
+++ b/app/src/main/java/com/amaze/filemanager/asynchronous/asynctasks/GenerateHashesTask.java
@@ -75,7 +75,7 @@ public class GenerateHashesTask extends AsyncTask<Void, String, String[]> {
       if (file.isSftp()) {
         md5 =
             SshClientUtils.execute(
-                new SshClientSessionTemplate(file.getPath()) {
+                new SshClientSessionTemplate<String>(file.getPath()) {
                   @Override
                   public String execute(Session session) throws IOException {
                     Session.Command cmd =
@@ -94,7 +94,7 @@ public class GenerateHashesTask extends AsyncTask<Void, String, String[]> {
                 });
         sha256 =
             SshClientUtils.execute(
-                new SshClientSessionTemplate(file.getPath()) {
+                new SshClientSessionTemplate<String>(file.getPath()) {
                   @Override
                   public String execute(Session session) throws IOException {
                     Session.Command cmd =
@@ -102,8 +102,7 @@ public class GenerateHashesTask extends AsyncTask<Void, String, String[]> {
                             String.format(
                                 "sha256sum -b \"%s\" | cut -c -64",
                                 SshClientUtils.extractRemotePathFrom(file.getPath())));
-                    String result =
-                        new String(IOUtils.readFully(cmd.getInputStream()).toByteArray());
+                    String result = IOUtils.readFully(cmd.getInputStream()).toString();
                     cmd.close();
                     if (cmd.getExitStatus() == 0) return result;
                     else {

--- a/app/src/main/java/com/amaze/filemanager/filesystem/HybridFile.java
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/HybridFile.java
@@ -219,13 +219,20 @@ public class HybridFile {
   public long lastModified() {
     switch (mode) {
       case SFTP:
-        return SshClientUtils.<Long>execute(
-            new SFtpClientTemplate(path) {
-              @Override
-              public Long execute(@NonNull SFTPClient client) throws IOException {
-                return client.mtime(SshClientUtils.extractRemotePathFrom(path));
-              }
-            });
+        final Long returnValue =
+            SshClientUtils.execute(
+                new SFtpClientTemplate<Long>(path) {
+                  @Override
+                  public Long execute(@NonNull SFTPClient client) throws IOException {
+                    return client.mtime(SshClientUtils.extractRemotePathFrom(path));
+                  }
+                });
+
+        if (returnValue == null) {
+          Log.e(TAG, "Error obtaining last modification time over SFTP");
+        }
+
+        return returnValue == null ? 0L : returnValue;
       case SMB:
         SmbFile smbFile = getSmbFile();
         if (smbFile != null) {
@@ -461,24 +468,32 @@ public class HybridFile {
     boolean isDirectory;
     switch (mode) {
       case SFTP:
-        return SshClientUtils.execute(
-            new SFtpClientTemplate(path) {
-              @Override
-              public Boolean execute(SFTPClient client) throws IOException {
-                try {
-                  return client
-                      .stat(SshClientUtils.extractRemotePathFrom(path))
-                      .getType()
-                      .equals(FileMode.Type.DIRECTORY);
-                } catch (SFTPException notFound) {
-                  Log.e(
-                      getClass().getSimpleName(),
-                      "Fail to execute isDirectory for SFTP path :" + path);
-                  notFound.printStackTrace();
-                  return false;
-                }
-              }
-            });
+        final Boolean returnValue =
+            SshClientUtils.<Boolean>execute(
+                new SFtpClientTemplate<Boolean>(path) {
+                  @Override
+                  public Boolean execute(SFTPClient client) throws IOException {
+                    try {
+                      return client
+                          .stat(SshClientUtils.extractRemotePathFrom(path))
+                          .getType()
+                          .equals(FileMode.Type.DIRECTORY);
+                    } catch (SFTPException notFound) {
+                      Log.e(
+                          getClass().getSimpleName(),
+                          "Fail to execute isDirectory for SFTP path :" + path);
+                      notFound.printStackTrace();
+                      return false;
+                    }
+                  }
+                });
+
+        if (returnValue == null) {
+          Log.e(TAG, "Error obtaining if path is directory over SFTP");
+        }
+
+        //noinspection SimplifiableConditionalExpression
+        return returnValue == null ? false : returnValue;
       case SMB:
         try {
           isDirectory =
@@ -571,13 +586,20 @@ public class HybridFile {
 
     switch (mode) {
       case SFTP:
-        return SshClientUtils.execute(
-            new SFtpClientTemplate(path) {
-              @Override
-              public Long execute(SFTPClient client) throws IOException {
-                return client.size(SshClientUtils.extractRemotePathFrom(path));
-              }
-            });
+        final Long returnValue =
+            SshClientUtils.<Long>execute(
+                new SFtpClientTemplate<Long>(path) {
+                  @Override
+                  public Long execute(SFTPClient client) throws IOException {
+                    return client.size(SshClientUtils.extractRemotePathFrom(path));
+                  }
+                });
+
+        if (returnValue == null) {
+          Log.e(TAG, "Error obtaining size of folder over SFTP");
+        }
+
+        return returnValue == null ? 0L : returnValue;
       case SMB:
         SmbFile smbFile = getSmbFile();
         size = (smbFile != null) ? FileUtils.folderSize(smbFile) : 0L;
@@ -631,9 +653,9 @@ public class HybridFile {
         size = spaceAllocation.getTotal() - spaceAllocation.getUsed();
         break;
       case SFTP:
-        size =
-            SshClientUtils.execute(
-                new SFtpClientTemplate(path) {
+        final Long returnValue =
+            SshClientUtils.<Long>execute(
+                new SFtpClientTemplate<Long>(path) {
                   @Override
                   public Long execute(@NonNull SFTPClient client) throws IOException {
                     try {
@@ -656,6 +678,12 @@ public class HybridFile {
                     }
                   }
                 });
+
+        if (returnValue == null) {
+          Log.e(TAG, "Error obtaining usable space over SFTP");
+        }
+
+        size = returnValue == null ? 0L : returnValue;
         break;
       case OTG:
         // TODO: Get free space from OTG when {@link DocumentFile} API adds support
@@ -689,9 +717,9 @@ public class HybridFile {
         size = spaceAllocation.getTotal();
         break;
       case SFTP:
-        size =
-            SshClientUtils.execute(
-                new SFtpClientTemplate(path) {
+        final Long returnValue =
+            SshClientUtils.<Long>execute(
+                new SFtpClientTemplate<Long>(path) {
                   @Override
                   public Long execute(@NonNull SFTPClient client) throws IOException {
                     try {
@@ -714,6 +742,12 @@ public class HybridFile {
                     }
                   }
                 });
+
+        if (returnValue == null) {
+          Log.e(TAG, "Error obtaining total space over SFTP");
+        }
+
+        size = returnValue == null ? 0L : returnValue;
         break;
       case OTG:
         // TODO: Find total storage space of OTG when {@link DocumentFile} API adds support
@@ -728,39 +762,35 @@ public class HybridFile {
   public void forEachChildrenFile(Context context, boolean isRoot, OnFileFound onFileFound) {
     switch (mode) {
       case SFTP:
-        try {
-          SshClientUtils.<Boolean>execute(
-              new SFtpClientTemplate(path) {
-                @Override
-                public Boolean execute(SFTPClient client) {
-                  try {
-                    for (RemoteResourceInfo info :
-                        client.ls(SshClientUtils.extractRemotePathFrom(path))) {
-                      boolean isDirectory = false;
-                      try {
-                        isDirectory = SshClientUtils.isDirectory(client, info);
-                      } catch (IOException ifBrokenSymlink) {
-                        Log.w(TAG, "IOException checking isDirectory(): " + info.getPath());
-                        continue;
-                      }
-                      HybridFileParcelable f = new HybridFileParcelable(path, isDirectory, info);
-                      onFileFound.onFileFound(f);
+        SshClientUtils.<Boolean>execute(
+            new SFtpClientTemplate<Boolean>(path) {
+              @Override
+              public Boolean execute(SFTPClient client) {
+                try {
+                  for (RemoteResourceInfo info :
+                      client.ls(SshClientUtils.extractRemotePathFrom(path))) {
+                    boolean isDirectory = false;
+                    try {
+                      isDirectory = SshClientUtils.isDirectory(client, info);
+                    } catch (IOException ifBrokenSymlink) {
+                      Log.w(TAG, "IOException checking isDirectory(): " + info.getPath());
+                      continue;
                     }
-                  } catch (IOException e) {
-                    Log.w("DEBUG.listFiles", "IOException", e);
-                    AppConfig.toast(
-                        context,
-                        context.getString(
-                            R.string.cannot_read_directory,
-                            parseAndFormatUriForDisplay(path),
-                            e.getMessage()));
+                    HybridFileParcelable f = new HybridFileParcelable(path, isDirectory, info);
+                    onFileFound.onFileFound(f);
                   }
-                  return true;
+                } catch (IOException e) {
+                  Log.w("DEBUG.listFiles", "IOException", e);
+                  AppConfig.toast(
+                      context,
+                      context.getString(
+                          R.string.cannot_read_directory,
+                          parseAndFormatUriForDisplay(path),
+                          e.getMessage()));
                 }
-              });
-        } catch (Exception e) {
-          e.printStackTrace();
-        }
+                return true;
+              }
+            });
         break;
       case SMB:
         try {
@@ -817,38 +847,31 @@ public class HybridFile {
     ArrayList<HybridFileParcelable> arrayList = new ArrayList<>();
     switch (mode) {
       case SFTP:
-        try {
-          arrayList =
-              SshClientUtils.execute(
-                  new SFtpClientTemplate(path) {
-                    @Override
-                    public ArrayList<HybridFileParcelable> execute(SFTPClient client) {
-                      ArrayList<HybridFileParcelable> retval =
-                          new ArrayList<HybridFileParcelable>();
-                      try {
-                        for (RemoteResourceInfo info :
-                            client.ls(SshClientUtils.extractRemotePathFrom(path))) {
-                          boolean isDirectory = false;
-                          try {
-                            isDirectory = SshClientUtils.isDirectory(client, info);
-                          } catch (IOException ifBrokenSymlink) {
-                            Log.w(TAG, "IOException checking isDirectory(): " + info.getPath());
-                            continue;
-                          }
-                          HybridFileParcelable f =
-                              new HybridFileParcelable(path, isDirectory, info);
-                          retval.add(f);
+        arrayList =
+            SshClientUtils.execute(
+                new SFtpClientTemplate<ArrayList<HybridFileParcelable>>(path) {
+                  @Override
+                  public ArrayList<HybridFileParcelable> execute(SFTPClient client) {
+                    ArrayList<HybridFileParcelable> retval = new ArrayList<>();
+                    try {
+                      for (RemoteResourceInfo info :
+                          client.ls(SshClientUtils.extractRemotePathFrom(path))) {
+                        boolean isDirectory = false;
+                        try {
+                          isDirectory = SshClientUtils.isDirectory(client, info);
+                        } catch (IOException ifBrokenSymlink) {
+                          Log.w(TAG, "IOException checking isDirectory(): " + info.getPath());
+                          continue;
                         }
-                      } catch (IOException e) {
-                        Log.w("DEBUG.listFiles", "IOException", e);
+                        HybridFileParcelable f = new HybridFileParcelable(path, isDirectory, info);
+                        retval.add(f);
                       }
-                      return retval;
+                    } catch (IOException e) {
+                      Log.w("DEBUG.listFiles", "IOException", e);
                     }
-                  });
-        } catch (Exception e) {
-          e.printStackTrace();
-          arrayList.clear();
-        }
+                    return retval;
+                  }
+                });
         break;
       case SMB:
         try {
@@ -900,11 +923,12 @@ public class HybridFile {
    *
    * @deprecated use {@link #getInputStream(Context)} which allows handling content resolver
    */
+  @Nullable
   public InputStream getInputStream() {
     InputStream inputStream;
     if (isSftp()) {
       return SshClientUtils.execute(
-          new SFtpClientTemplate(path) {
+          new SFtpClientTemplate<InputStream>(path) {
             @Override
             public InputStream execute(SFTPClient client) throws IOException {
               final RemoteFile rf = client.open(SshClientUtils.extractRemotePathFrom(path));
@@ -938,6 +962,7 @@ public class HybridFile {
     return inputStream;
   }
 
+  @Nullable
   public InputStream getInputStream(Context context) {
     InputStream inputStream;
 
@@ -945,7 +970,7 @@ public class HybridFile {
       case SFTP:
         inputStream =
             SshClientUtils.execute(
-                new SFtpClientTemplate(path, false) {
+                new SFtpClientTemplate<InputStream>(path, false) {
                   @Override
                   public InputStream execute(final SFTPClient client) throws IOException {
                     final RemoteFile rf = client.open(SshClientUtils.extractRemotePathFrom(path));
@@ -1010,12 +1035,13 @@ public class HybridFile {
     return inputStream;
   }
 
+  @Nullable
   public OutputStream getOutputStream(Context context) {
     OutputStream outputStream;
     switch (mode) {
       case SFTP:
         return SshClientUtils.execute(
-            new SshClientTemplate(path, false) {
+            new SshClientTemplate<OutputStream>(path, false) {
               @Override
               public OutputStream execute(final SSHClient ssh) throws IOException {
                 final SFTPClient client = ssh.newSFTPClient();
@@ -1074,9 +1100,9 @@ public class HybridFile {
   public boolean exists() {
     boolean exists = false;
     if (isSftp()) {
-      exists =
-          SshClientUtils.execute(
-              new SFtpClientTemplate(path) {
+      final Boolean executionReturn =
+          SshClientUtils.<Boolean>execute(
+              new SFtpClientTemplate<Boolean>(path) {
                 @Override
                 public Boolean execute(SFTPClient client) throws IOException {
                   try {
@@ -1086,6 +1112,13 @@ public class HybridFile {
                   }
                 }
               });
+
+      if (executionReturn == null) {
+        Log.e(TAG, "Error obtaining existance of file over SFTP");
+      }
+
+      //noinspection SimplifiableConditionalExpression
+      exists = executionReturn == null ? false : executionReturn;
     } else if (isSmb()) {
       try {
         SmbFile smbFile = getSmbFile(2000);
@@ -1163,13 +1196,13 @@ public class HybridFile {
   public void mkdir(Context context) {
     if (isSftp()) {
       SshClientUtils.execute(
-          new SFtpClientTemplate(path) {
+          new SFtpClientTemplate<Void>(path) {
             @Override
-            public Void execute(SFTPClient client) {
+            public Void execute(@NonNull SFTPClient client) {
               try {
                 client.mkdir(SshClientUtils.extractRemotePathFrom(path));
               } catch (IOException e) {
-                e.printStackTrace();
+                Log.e(TAG, "Error making directory over SFTP", e);
               }
               // FIXME: anything better than throwing a null to make Rx happy?
               return null;

--- a/app/src/main/java/com/amaze/filemanager/filesystem/Operations.java
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/Operations.java
@@ -429,9 +429,9 @@ public class Operations {
           return null;
         } else if (oldFile.isSftp()) {
           SshClientUtils.execute(
-              new SFtpClientTemplate(oldFile.getPath()) {
+              new SFtpClientTemplate<Void>(oldFile.getPath()) {
                 @Override
-                public <Void> Void execute(@NonNull SFTPClient client) {
+                public Void execute(@NonNull SFTPClient client) {
                   try {
                     client.rename(
                         SshClientUtils.extractRemotePathFrom(oldFile.getPath()),

--- a/app/src/main/java/com/amaze/filemanager/filesystem/cloud/CloudUtil.java
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/cloud/CloudUtil.java
@@ -57,6 +57,7 @@ import android.net.Uri;
 import android.os.AsyncTask;
 import android.widget.Toast;
 
+import androidx.annotation.Nullable;
 import androidx.documentfile.provider.DocumentFile;
 
 import net.schmizz.sshj.sftp.RemoteFile;
@@ -245,6 +246,7 @@ public class CloudUtil {
   }
 
   /** Get an input stream for thumbnail for a given {@link IconDataParcelable} */
+  @Nullable
   public static InputStream getThumbnailInputStreamForCloud(Context context, String path) {
     InputStream inputStream;
     HybridFile hybridFile = new HybridFile(OpenMode.UNKNOWN, path);
@@ -255,7 +257,7 @@ public class CloudUtil {
       case SFTP:
         inputStream =
             SshClientUtils.execute(
-                new SFtpClientTemplate(hybridFile.getPath(), false) {
+                new SFtpClientTemplate<InputStream>(hybridFile.getPath(), false) {
                   @Override
                   public InputStream execute(final SFTPClient client) throws IOException {
                     final RemoteFile rf =

--- a/app/src/main/java/com/amaze/filemanager/filesystem/ssh/SFtpClientTemplate.kt
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/ssh/SFtpClientTemplate.kt
@@ -27,7 +27,7 @@ import java.io.IOException
  * Template class for executing actions with [SFTPClient] while leave the complexities of
  * handling connection and session setup/teardown to [SshClientUtils].
  */
-abstract class SFtpClientTemplate
+abstract class SFtpClientTemplate <T>
 /**
  * If closeClientOnFinish is set to true, calling code needs to handle closing of [ ] session.
  *
@@ -46,5 +46,5 @@ constructor(@JvmField val url: String, @JvmField val closeClientOnFinish: Boolea
      * @return Result of the execution of the type requested
      */
     @Throws(IOException::class)
-    abstract fun <T> execute(client: SFTPClient): T
+    abstract fun execute(client: SFTPClient): T
 }

--- a/app/src/main/java/com/amaze/filemanager/filesystem/ssh/SshClientSessionTemplate.kt
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/ssh/SshClientSessionTemplate.kt
@@ -23,7 +23,7 @@ package com.amaze.filemanager.filesystem.ssh
 import net.schmizz.sshj.connection.channel.direct.Session
 import java.io.IOException
 
-abstract class SshClientSessionTemplate
+abstract class SshClientSessionTemplate <T>
 /**
  * Constructor.
  *
@@ -41,5 +41,5 @@ abstract class SshClientSessionTemplate
      * @return Result of the execution of the type requested
      */
     @Throws(IOException::class)
-    abstract fun <T> execute(sshClientSession: Session): T
+    abstract fun execute(sshClientSession: Session): T
 }

--- a/app/src/main/java/com/amaze/filemanager/filesystem/ssh/SshClientTemplate.kt
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/ssh/SshClientTemplate.kt
@@ -20,6 +20,7 @@
 
 package com.amaze.filemanager.filesystem.ssh
 
+import androidx.annotation.Nullable
 import net.schmizz.sshj.SSHClient
 import java.io.IOException
 
@@ -27,7 +28,7 @@ import java.io.IOException
  * Template class for executing actions with [SSHClient] while leave the complexities of
  * handling connection setup/teardown to [SshClientUtils].
  */
-abstract class SshClientTemplate
+abstract class SshClientTemplate <T>
 /**
  * Constructor, with closeClientOnFinish set to true (that the connection must close after `
  * execute`.
@@ -47,14 +48,6 @@ constructor(@JvmField val url: String, @JvmField val closeClientOnFinish: Boolea
      * @return Result of the execution of the type requested
      </T> */
     @Throws(IOException::class)
-    abstract fun <T> execute(client: SSHClient): T
-    /**
-     * Constructor.
-     *
-     * @param url SSH connection URL, in the form of `
-     * ssh://<username>:<password>@<host>:<port>` or `
-     * ssh://<username>@<host>:<port>`
-     * @param closeClientOnFinish if set to false, connection will be kept on after `execute
-     ` *  finishes. This is useful for other calls to reuse the connection.
-     */
+    @Nullable
+    abstract fun execute(client: SSHClient): T
 }

--- a/app/src/main/java/com/amaze/filemanager/filesystem/ssh/SshClientUtils.java
+++ b/app/src/main/java/com/amaze/filemanager/filesystem/ssh/SshClientUtils.java
@@ -75,7 +75,7 @@ public abstract class SshClientUtils {
    * @param <T> Type of return value
    * @return Template execution results
    */
-  public static final <T> T execute(@NonNull SshClientTemplate template) {
+  public static <T> T execute(@NonNull SshClientTemplate<T> template) {
     SSHClient client = SshConnectionPool.INSTANCE.getConnection(extractBaseUriFrom(template.url));
     if (client == null) {
       client = SshConnectionPool.INSTANCE.getConnection(template.url);
@@ -106,9 +106,9 @@ public abstract class SshClientUtils {
    * @param <T> Type of return value
    * @return Template execution results
    */
-  public static final <T> T execute(@NonNull final SshClientSessionTemplate template) {
+  public static <T> T execute(@NonNull final SshClientSessionTemplate<T> template) {
     return execute(
-        new SshClientTemplate(template.url, false) {
+        new SshClientTemplate<T>(template.url, false) {
           @Override
           public T execute(SSHClient client) {
             Session session = null;
@@ -139,10 +139,12 @@ public abstract class SshClientUtils {
    * @param <T> Type of return value
    * @return Template execution results
    */
-  public static final <T> T execute(@NonNull final SFtpClientTemplate template) {
-    return execute(
-        new SshClientTemplate(template.url, false) {
+  @Nullable
+  public static <T> T execute(@NonNull final SFtpClientTemplate<T> template) {
+    final SshClientTemplate<T> sshClient =
+        new SshClientTemplate<T>(template.url, false) {
           @Override
+          @Nullable
           public T execute(SSHClient client) {
             SFTPClient sftpClient = null;
             T retval = null;
@@ -162,7 +164,9 @@ public abstract class SshClientUtils {
             }
             return retval;
           }
-        });
+        };
+
+    return execute(sshClient);
   }
 
   /**
@@ -172,7 +176,7 @@ public abstract class SshClientUtils {
    * @param fullUri SSH URL
    * @return SSH URL with the password (if exists) encrypted
    */
-  public static final String encryptSshPathAsNecessary(@NonNull String fullUri) {
+  public static String encryptSshPathAsNecessary(@NonNull String fullUri) {
     String uriWithoutProtocol =
         fullUri.substring(SSH_URI_PREFIX.length(), fullUri.lastIndexOf('@'));
     try {
@@ -192,7 +196,7 @@ public abstract class SshClientUtils {
    * @param fullUri SSH URL
    * @return SSH URL with the password (if exists) decrypted
    */
-  public static final String decryptSshPathAsNecessary(@NonNull String fullUri) {
+  public static String decryptSshPathAsNecessary(@NonNull String fullUri) {
     String uriWithoutProtocol =
         fullUri.substring(SSH_URI_PREFIX.length(), fullUri.lastIndexOf('@'));
     try {
@@ -214,7 +218,7 @@ public abstract class SshClientUtils {
    * @param fullUri Full SSH URL
    * @return The remote path part of the full SSH URL
    */
-  public static final String extractBaseUriFrom(@NonNull String fullUri) {
+  public static String extractBaseUriFrom(@NonNull String fullUri) {
     String uriWithoutProtocol = fullUri.substring(SSH_URI_PREFIX.length());
     return uriWithoutProtocol.indexOf('/') == -1
         ? fullUri
@@ -230,7 +234,7 @@ public abstract class SshClientUtils {
    * @param fullUri Full SSH URL
    * @return The remote path part of the full SSH URL
    */
-  public static final String extractRemotePathFrom(@NonNull String fullUri) {
+  public static String extractRemotePathFrom(@NonNull String fullUri) {
     String uriWithoutProtocol = fullUri.substring(SSH_URI_PREFIX.length());
     return uriWithoutProtocol.indexOf('/') == -1
         ? "/"
@@ -243,7 +247,7 @@ public abstract class SshClientUtils {
    *
    * @param client {@link SSHClient} instance
    */
-  public static final void tryDisconnect(SSHClient client) {
+  public static void tryDisconnect(SSHClient client) {
     if (client != null && client.isConnected()) {
       try {
         client.disconnect();
@@ -326,10 +330,10 @@ public abstract class SshClientUtils {
   }
 
   public static @FolderState int checkFolder(@NonNull String path) {
-    return Single.<Integer>fromCallable(
+    return Single.fromCallable(
             () ->
                 execute(
-                    new SFtpClientTemplate(extractBaseUriFrom(path)) {
+                    new SFtpClientTemplate<Integer>(extractBaseUriFrom(path)) {
                       @Override
                       public @FolderState Integer execute(@NonNull SFTPClient client)
                           throws IOException {

